aspnetcore/mvc/advanced/custom-model-binding.md

```diff
@@ -125,6 +125,87 @@ Adding your provider to the end of the collection may result in a built-in model
 
 [!code-csharp[](custom-model-binding/sample/CustomModelBindingSample/Startup.cs?name=callout&highlight=5-9)]
 
+### Polymorphic model binding
+
+One possible reason to do custom model binding is to use some value from the request to produce different types of models. We would generally recommend avoiding this pattern since it makes it non-trivial to reason about your bound models. However, if your application requires such a pattern, a pattern for implementing it would look like this:
+
+```C#
+[ModelBinder(typeof(DeviceModelBinder))]
+public class Device
+{
+    public string Kind { get; set; }
+}
+
+public class Laptop : Device
+{
+    public string CPUIndex { get; set; }
+}
+
+public class SmartPhone : Device
+{
+    public string ScreenSize { get; set; }
+}
+
+public class DeviceModelBinder : IModelBinder
+{
+    public Task BindModelAsync(ModelBindingContext bindingContext)
+    {
+        var name = ModelNames.CreatePropertyModelName(
+            bindingContext.ModelName,
+            nameof(Device.Kind));
+
+        var deviceKind = bindingContext.ValueProvider.GetValue(name).FirstValue;
+
+        if (string.IsNullOrEmpty(deviceKind))
+        {
+            bindingContext.Result = ModelBindingResult.Failed();
+            return Task.CompletedTask;
+        }
+
+        Device result;
+        if (deviceKind == "Laptop")
+        {
+            var modelName = ModelNames.CreatePropertyModelName(bindingContext.ModelName, nameof(Laptop.CPUIndex));
+            var cpuIndex = bindingContext.ValueProvider.GetValue(modelName).FirstValue;
```

Comment:
@rynowak Did you mean `ComplexTypeModelBinder`? I can have a go at an implementation that subclasses that if it is what you meant.

Reply:
@pranavkm can you confirm `ComplexTypeModelBinder`

Reply:
> I think we've done this in the past by subclassing ComplexObjectModelBinder right?

yeah, we could do it. I was trying to avoid using it just because it's kinda gross to set it up.

Reply:
> yeah, we could do it. I was trying to avoid using it just because it's kinda gross to set it up.

@serpent5  is working on this in #14237. Should he stop? Should he provide both approaches?

Reply:
No, it's fine.